upload-api/tables/delegations.js

```diff
@@ -0,0 +1,161 @@
+import { base32 } from 'multiformats/bases/base32'
+import {
+  DynamoDBClient,
+  QueryCommand,
+  BatchWriteItemCommand,
+  DescribeTableCommand,
+} from '@aws-sdk/client-dynamodb'
+import { marshall, unmarshall } from '@aws-sdk/util-dynamodb'
+import * as Ucanto from '@ucanto/interface'
+
+import { CID } from 'multiformats/cid'
+import {
+  bytesToDelegations,
+  delegationsToBytes
+} from '@web3-storage/access/encoding'
+
+/**
+ * @typedef {Ucanto.Delegation} Delegation
+ */
+
+/**
+ * Abstraction layer to handle operations on Store Table.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {import('../types').DelegationsBucket} bucket
+ * @param {object} [options]
+ * @param {string} [options.endpoint]
+ */
+export function createDelegationsTable (region, tableName, bucket, options = {}) {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint,
+  })
+
+  return useDelegationsTable(dynamoDb, tableName, bucket)
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} tableName
+ * @param {import('../types').DelegationsBucket} bucket
+ * @returns {import('@web3-storage/upload-api').DelegationsStorage}
+ */
+export function useDelegationsTable (dynamoDb, tableName, bucket) {
+  return {
+    putMany: async (...delegations) => {
+      if (delegations.length === 0) {
+        return {
+          ok: {}
+        }
+      }
+      await writeDelegations(bucket, delegations)
+      // TODO: we should look at the return value of this BatchWriteItemCommand and either retry or clean up delegations that we fail to index
```

Comment:
I guess that for metrics might be good to have this output in the receipt for UCAN invocation. What do you think?

Reply:
hm I'm not sure I follow... are you saying that you think it would be good to record when this fails in the receipt for the invocation of the UCAN that kicked off this write? sounds right to me - I think all we need to do is return an appropriate error when this fails, is that right?

worth noting that this code will be much simpler if we stop saving delegations to a separate bucket!